set resource requests and limits on tank object

josibake opened the pull request:
tell k8s up front the min and max CPU, memory we want for a tank. can be overriden from a graph file if a tank needs more/less.

did some testing, and this allows k8s to do autoscaling when trying to start a network. this means if a cluster does not have enough nodes and you try to start a network with 50 tanks, k8s will keep the pods in pending until it adds more nodes to the cluster. this allows for much more graceful cluster management.

annoyingly, if the pod is already scheduled and the workload of the pod increases, k8s can't do cluster autoscaling anymore, but there might be another way to tackle this via vertical scaling. as of today, the best way to handle this would be to set a much higher `limits`, which should help? (not tested).

## testing

tested on digital ocean cluster with autoscaling, has not been tested on Minikube or docker desktop, but _should_ work since requests and limits are not k8s specific config options.

m3dwards commented:
Ran on Docker Desktop's Kubernetes and could see:

```shell
"resources": {
    "limits": {
        "cpu": "1",
        "memory": "1500Mi"
    },
    "requests": {
        "cpu": "500m",
        "memory": "500Mi"
    }
},
```

m3dwards commented:
CI failed, perhaps because minikube doesn't think it has enough resources to start the tanks

willcl-ark commented:
Added a commit to
- fix schema
- validate generated graphs during creation
- apply default resource limits to all created graphs (@josibake do we want this?)

willcl-ark commented:
An alternative using user-friendly "profiles": https://github.com/willcl-ark/warnet/pull/24

pinheadmz commented on src/warnet/utils.py:
```diff
@@ -380,7 +380,9 @@ def set_execute_permission(file_path):
 
-def create_cycle_graph(n: int, version: str, bitcoin_conf: str | None, random_version: bool):
+def create_cycle_graph(
+    n: int, version: str, bitcoin_conf: str | None, random_version: bool, resources: dict | None
```
`resources` isnt being used in the function?

pinheadmz reviewed (commented):
Can't we just set lower resource usage in the CI so the tests can run as written without any changes?

pinheadmz commented:
Error on docker desktop, `warcli network start` (so, default.graphml)

```
2024-08-20 12:04:09 | DEBUG   | k8s      | Deploying pods
2024-08-20 12:04:09 | DEBUG   | k8s      | Creating bitcoind container for tank 0
2024-08-20 12:04:09 | ERROR   | server   | Unhandled exception starting warnet: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "/root/warnet/src/warnet/server.py", line 469, in thread_start
    wn.warnet_up()
  File "/root/warnet/src/warnet/warnet.py", line 176, in warnet_up
    self.container_interface.up(self)
  File "/root/warnet/src/warnet/backend/kubernetes_backend.py", line 57, in up
    self.deploy_pods(warnet)
  File "/root/warnet/src/warnet/backend/kubernetes_backend.py", line 682, in deploy_pods
    bitcoind_container = self.create_bitcoind_container(tank)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/warnet/src/warnet/backend/kubernetes_backend.py", line 410, in create_bitcoind_container
    requests=tank.resources["requests"],
             ~~~~~~~~~~~~~~^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable


```

pinheadmz commented:
I *think* the bug Im getting is in the default set in schema.json

pinheadmz commented:
example fix:

```diff
diff --git a/src/warnet/graph_schema.json b/src/warnet/graph_schema.json
index aa1518f..4b178a4 100644
--- a/src/warnet/graph_schema.json
+++ b/src/warnet/graph_schema.json
@@ -69,25 +69,13 @@
         "comment": "A string of arguments for the lightning network node in command-line format, e.g. '--protocol.wumbo-channels --bitcoin.timelockdelta=80'"
       },
       "resources": {
-            "type": "object",
-            "comment": "Kubernetes resource requests and limits for the node",
-            "properties": {
-              "requests": {
-                "type": "object",
-                "properties": {
-                  "cpu": {"type": "string", "default": "500m"},
-                  "memory": {"type": "string", "default": "500Mi"}
-                }
-              },
-              "limits": {
-                "type": "object",
-                "properties": {
-                  "cpu": {"type": "string", "default":"1000m"},
-                  "memory": {"type": "string", "default":"1500Mi"}
-                }
-              }
-            }
+        "type": "object",
+        "comment": "Kubernetes resource requests and limits for the node",
+        "default": {
+          "requests": {"cpu": "100m", "memory": "250Mi"},
+          "limits": {"cpu": "250m", "memory": "500Mi"}
         }
+      }
     },
     "additionalProperties": false,
     "oneOf": [
```

josibake committed:
ci: increase minikube resources

now that we are explicitly requesting resources,
pods will not get scheduled unless the cluster
has enough resources to satisfy the initial request.

this was causing our current tests to fail, so bump
minikubes resources. the tests will be scaled down
in the next commits

josibake committed:
tests: slim down 12_node_ring to 6 nodes

josibake committed:
tests: slim down ten_semi to six_semi

josibake committed:
tests: slim down connect dag from 10 to 6

josibake committed:
tests: slim down permutations.graphml

test with 6 nodes instead of 8.
also replace version with image: this sets us up for a later change
to remove version from the config completely

josibake committed:
tests: replace version with image

we use image for everything now and version is a more of a leftover
from before when we didnt have a docker registry set up. in a later
commit, we will completely remove version from the jsonschema file

josibake committed:
add resources key to schema file

josibake commented:
Closing as this is now obsolete. We can now configure everything in one place using the network.yaml/node-defaults.yaml and namespaces.yaml/namespace-defaults.yaml helm files.

Only thing currently missing is good documentation and examples on how to use the yaml configuration files which I think we will need before tabconf, opened a backlog issue for that here: https://github.com/bitcoin-dev-project/warnet/issues/496
